flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/BaumWelch.scala

```diff
@@ -0,0 +1,346 @@
+                     sizeOfObservationVocab: Int):
+  DataSet[(DenseVector, DenseMatrix, DenseMatrix)] = {
+    results.groupBy(0, 1, 2).sum(3).map(t => (t._1, t._2, List((t._3, t._4)), t._4))
+      .groupBy(0, 1).reduce((tuple1, tuple2) =>
+      (tuple1._1, tuple1._2, tuple1._3 ::: tuple2._3, tuple1._4 + tuple2._4))
+      .reduceGroup(iterator => {
+        var initialProbabilities: DenseVector = null
+        var transitionProbabilities: DenseMatrix = null
+        var emissionProbabilities: DenseMatrix = null
+        var count: Double = 0
+        iterator.foreach(p => {
+          if (p._1.equals("i")) {
+            if (initialProbabilities == null) {
+              initialProbabilities = new DenseVector(new Array[Double](numberOfStates))
+            }
+            initialProbabilities(p._2.toInt) = p._4
+            count += p._4
+          } else if (p._1.equals("t")) {
+            if (transitionProbabilities == null) {
+              transitionProbabilities = new DenseMatrix(numberOfStates, numberOfStates,
+                new Array[Double](numberOfStates * numberOfStates))
+            }
+            p._3.foreach(t => transitionProbabilities(p._2.toInt, t._1.toInt) = t._2 / p._4)
+          } else if (p._1.equals("e")) {
+            if (emissionProbabilities == null) {
+              emissionProbabilities = new DenseMatrix(numberOfStates, sizeOfObservationVocab,
+                new Array[Double](numberOfStates * sizeOfObservationVocab))
+            }
+            p._3.foreach(t => emissionProbabilities(p._2.toInt, t._1.toInt) = t._2 / p._4)
+          }
+        })
+
+        for (i <- 0 until initialProbabilities.size) {
+          initialProbabilities(i) = initialProbabilities(i) / count
+        }
+        (initialProbabilities, transitionProbabilities, emissionProbabilities)
+      })
+  }
+
+  /**
+   * forward prodecure of the Baum-Welch algorithm
+   *
+   * @param initialProbabilities initial probabilities
+   * @param transitionProbabilities transition probabilities
+   * @param emissionProbabilities emission probabilities
+   * @param numberOfStates number of states
+   * @param observations training data
+   * @return forward probabilities
+   */
+  def forward(initialProbabilities: DenseVector, transitionProbabilities: DenseMatrix,
+              emissionProbabilities: DenseMatrix, numberOfStates: Int,
+              observations: Array[Int]): Array[Array[Double]] = {
+    val f: Array[Array[Double]] = Array.ofDim[Double](numberOfStates, observations.length)
+
+    //initial probabilities
+    for (state <- 0 until f.length) {
+      f(state)(0) = initialProbabilities(state) * emissionProbabilities(state, observations(0))
+    }
+
+    for (o <- 0 until observations.length - 1) {
+      for (state1 <- 0 until numberOfStates) {
+        for (state2 <- 0 until numberOfStates) {
+          f(state1)(o + 1) += f(state2)(o) * transitionProbabilities(state1, state2) *
```

Comment:
Does `transitionProbabilities(state1, state2)` say `Pr(X_t + 1 = state2 | X_t = state1)` for an arbitrary `t`?

If this is the case and assuming you're right, then Wikipedia is wrong (https://en.wikipedia.org/wiki/Baum%E2%80%93Welch_algorithm). 

Maybe I'm totally wrong here, since the last time I did the math for Baum-Welch is 2 years ago, but I think that the transition matrix is not necessarily symmetric and thus you cannot write `transitionProbabilities(state1, state2)`. Maybe you can help me with the math here (derivation).